### PR TITLE
README Update: Minor Format Changes and new Code Usage Part

### PR DESCRIPTION
**Pull request by danrgll**
Updated the README with minor formatting changes for better readability and introduced a new, more example oriented 'Code Usage' section.

### CODE REVIEW

File: docs/analyse.md
Comment (eddiebergman) [inline docs/analyse.md:86]: > [!TIP]
> The docs used here are `material-mkdocs` which has a lot of nice little features.
> One of them is called [_admonations_](https://squidfunk.github.io/mkdocs-material/reference/admonitions/).

> [!NOTE]
> Github markdown a different syntax but the idea is the same

File: docs/analyse.md
Comment (eddiebergman) [inline docs/analyse.md:121]: This really seems like `API` documentation that needs to be automated as this is definitely going to go out of date and be incorrect soon. Not an issue for this PR but just want to note it.

File: docs/analyse.md
Comment (eddiebergman) [inline docs/analyse.md:123]: This section here is likely better to live in some `logging_tensorboard_example.py` and then importing it into the markdown with [this](https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#embedding-external-files).

````
```title="Logging with tensorboard in NePS"
--8<-- "path/to/logging_tensorboard_example.py"
```
````

File: mkdocs.yml
Comment (eddiebergman) [inline mkdocs.yml:21]: nice

**Review by eddiebergman — APPROVED** [review]
Nothing really holding it back, although a few things could be fixed up.

Reviewing this makes me realise NePS really needs automated API documentation. I can set that up at some point so you can reference it whereever, i.e.

```md
### Some page about logging
We can log, here's the signature

::: neps.logging.tensorboard.Thing
    options:
    	show_members: False
```    	
The above thing will render with mkdocs and pull in the api for `Thing` and you can control what it shows. That means it won't go out of date with the docs.

> [!TIP]
> You can make [_admonations_](https://squidfunk.github.io/mkdocs-material/reference/admonitions/#admonitions) with `material-mkdocs` that look kind of like this **Tip** block but with the following syntax
> ````md
> !!! note "Some Title"
>
> 	This is an expanded note
>	    
> ??? example "Some Other title"
>	
>	This is an example with code
>		
>	```python
>	print("yay")
>	```
> ````

Having a read through material docs documentation can be quite cool :)

https://squidfunk.github.io/mkdocs-material/reference/

File: docs/analyse.md
Comment (TarekAbouChakra) [reply docs/analyse.md:86]: Done, thanks for this.

File: docs/analyse.md
Comment (TarekAbouChakra) [reply docs/analyse.md]: Currently I did it so because it is more user friendly with less details than this one [here](https://github.com/automl/neps/blob/master/neps_examples/convenience/neps_tblogger_tutorial.py). However, I am open to changing the original example to the new one, knowing that the main purpose of the example is to showcase TensorBoard integration in NePS.

File: README.md
Comment (Neeratyoy) [inline README.md:103]: Once the citation block is added, can we merge this?

Comment (TarekAbouChakra) [reply README.md:103]: Citations are now added [here](https://github.com/automl/neps/blob/readme_changes_danrgll/docs/citations.md).

File: README.md
Comment (Neeratyoy) [inline README.md:1]: Can we move the `Alternatives` block as the last block please?

File: docs/analyse.md
Comment (Neeratyoy) [inline docs/analyse.md:313]: This feels way too much content on one page of documentation, especially something that the user HAS to scroll through.
Don't we have an example? Or we can move this as an example and link it here.

File: docs/analyse.md
Comment (Neeratyoy) [inline docs/analyse.md:64]: Firstly, do you think we can create a dedicated page for this and link it in the `Analyzing` page.

In any case, I feel this section should come up much higher in this page as this is a bigger USP for NePS than whatever is above it. So should be called out and emphasized more.
Also, this can also serve as an HPO search quality status check too! 

So calling it just `visualization` feels heavy underselling.

File: docs/analyse.md
Comment (Neeratyoy) [inline docs/analyse.md:346]: These are very cool additions.

One minor nitpick is that it is not super clear what the text highlights mean on the images.

Secondly, if we are putting up images on the doc, can we do a proper run with more parameter ranges and budgets, and then put the images from that run? 
That should also give us a good scatter plot image, and a more wholesome set of learning curves.

Right now, it really looks like a poor man's HPO run and can be misinterpreted as not able to handle scale! 😅

File: README.md
Comment (Neeratyoy) [inline README.md:1]: Overall, looks improved!

File: README.md
Comment (Neeratyoy) [inline README.md:12]: I shall not merge this yet.

This is a good first iteration to change the landing page. I do like elements of it.
However, the major reason why I shall not merge for now is consistency.

I think we can improve the user experience with every hyperlink directing the user to the same visual experience. That is, we either link everything to another GitHub page or only link to pages in the documentation.

Secondly, if we are putting 5 links right in the beginning, such that someone is more likely to click. The subsequent pages should have more content than what is now. 🙈 So I guess if we want to change the first interface, we can do so only after a little more backend content...

For point 1, either in the main point or as a subpoint, please mention these 2 papers as:
* [PriorBand: Practical Hyperparameter Optimization in the Age of Deep Learning, NeurIPS 2023](https://arxiv.org/abs/2306.12370)
* [πBO: Augmenting Acquisition Functions with User Beliefs for Bayesian Optimization, ICLR 2022](https://arxiv.org/abs/2204.11051)

For point 2, same with:
* [Construction of Hierarchical Neural Architecture Search Spaces based on Context-free Grammars, NeurIPS 2023](https://arxiv.org/abs/2211.01842)

File: README.md
Comment (Neeratyoy) [inline README.md:26]: If we do create a new page in our docs for this, maybe we can link the pytorch and lightning templates, and also the docs from yaml APIs. What do you think?
For the latter we can show both easy use for non-HPO folks and also for people who know some HPO and want to try different optimizers.

File: README.md
Comment (Neeratyoy) [inline README.md:8]:
```suggestion
Welcome to NePS, a powerful and flexible Python library for hyperparameter optimization (HPO) and neural architecture search (NAS) with its primary goal: enable HPO adoption in practice for deep learners!

NePS houses recently published and some more well-established algorithms that are all capable of being run massively parallel on any distributed setup, with tools to analyze runs, restart runs, etc.

Read our documentation and the following document for how to use NePS!

```

Comment (Neeratyoy) [reply README.md:8]: Is it possible to link our documentation right at the beginning, early on, where we write `documentation`?

File: README.md
Comment (Neeratyoy) [inline README.md:43]: Do we link the `templates` folder here?

File: README.md
Comment (Neeratyoy) [inline README.md:87]: Assuming we have more comprehensive documentation on how to construct or use neps search space, can we replace this example with a `yaml` file equivalent, just to show that the `basic usage` code requirement is quite low from `neps`? 

Of course, we can do this in the main README only if we also link a page where we can write the Pythonic version of the same thing and show more details such as other hyperparameter, use of priors, and grammars.

File: docs/analyse.md
Comment (Neeratyoy) [inline docs/analyse.md:73]: is it possible to do a ```python``` block here and show that this function can only be used when inside a `run_pipeline()`.
Of course, a dummy run_pipeline with `...` before and after the `tblogger.log` call.

**Comment by Neeratyoy** [comment]
As you might see, I borrowed some changes from this PR and updated the current README to make certain important changes to it for the release! 
This PR can now iterate for a better version for the next release :)

File: docs/analyse.md
Comment (TarekAbouChakra) [reply docs/analyse.md:313]: We do have an example. We could link the example here as the guide and showcase the images only.

File: docs/analyse.md
Comment (TarekAbouChakra) [reply docs/analyse.md]: I removed the text highlights, it seemed all in all a cleaner look not include them. Linked the example and specified the modifications needed to be done to obtain the visualization results that are now available.

Used a more sophisticated HPO method + Images seem to be more meaningful.

File: docs/analyse.md
Comment (TarekAbouChakra) [reply docs/analyse.md]: It is now higher in the page to show its importance, and has its own title under `TensorBoard Integration`.

File: README.md
Comment (TarekAbouChakra) [reply README.md:12]: Our doc is currently a work in progress. I would personally motivate us linking another GitHub page till we iterate through the docs later again and update this to link to documentation. What do you think ?

Comment (TarekAbouChakra) [reply README.md:17]: Papers are now mentioned. Have a look at the current landing page README [here](https://github.com/automl/neps/tree/readme_changes_danrgll)

**Pull request merged**
